### PR TITLE
[Cache] Add a Chain adapter

### PR DESCRIPTION
**Pull request by dunglas**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | n/a |
| License | MIT |
| Doc PR | todo |

Allows to chain adapters. Replaces #17452.

### CODE REVIEW

**Comment by Tobion** [comment]
So every PSR-6 cache implementation must implement its own ChainCache because each one can only work with his own CacheItem?

**Comment by dunglas** [comment]
You can use an external CacheItemPool if you wrap it with the ProxyApadpter before adding it to the chain.

**Comment by dunglas** [comment]
Btw it is possible to automatically wrap external pools in the constructor. I'll update the PR.

File: src/Symfony/Component/Cache/Adapter/ChainAdapter.php [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+class ChainAdapter implements AdapterInterface
+{
+    private $adapters = array();
+
+    /**
+     * @param AdapterInterface[] $adapters
+     */
+    public function __construct(array $adapters)
+    {
+        if (2 > count($adapters)) {
```
Comment (fabpot) [inline src/Symfony/Component/Cache/Adapter/ChainAdapter.php]: Why? When making this configurable by the user, you might end up with just one adapter, which is fine. I would change the check to `1 < count()`.

File: src/Symfony/Component/Cache/Adapter/ChainAdapter.php [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getItems(array $keys = array())
+    {
+        $items = array();
+        foreach ($keys as $key) {
+            $items[$key] = $this->getItem($key);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/ChainAdapter.php]: This implementation is naïve and prevents efficient retrial of multiple items. We should use getItems() on chained adapters instead, until all of the items are found

File: src/Symfony/Component/Cache/Adapter/AdapterInterface.php [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+namespace Symfony\Component\Cache\Adapter;
+
+use Psr\Cache\CacheItemPoolInterface;
+
+/**
+ * Marker interface for adapters managing {@see \Symfony\Component\Cache\CacheItem} instances.
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+interface AdapterInterface extends CacheItemPoolInterface
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/AdapterInterface.php]: imo this interface deserves a dedicated PR

Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/AdapterInterface.php]: in which ProxyAdapter is changed to use it to detect compatible CacheItems

**Comment by nicolas-grekas** [comment]
Status: needs work

File: src/Symfony/Component/Cache/Tests/Adapter/ChainAdapterTest.php [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+use Symfony\Component\Cache\Adapter\ChainAdapter;
+use Symfony\Component\Cache\Tests\Fixtures\ExternalAdapter;
+
+/**
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+class ChainAdapterTest extends CachePoolTest
+{
+    protected $skippedTests = array(
+        'testDeferredSaveWithoutCommit' => 'Assumes a shared cache which ArrayAdapter is not.',
```
Comment (stof) [inline src/Symfony/Component/Cache/Tests/Adapter/ChainAdapterTest.php]: this looks suspicious to me here

Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Tests/Adapter/ChainAdapterTest.php]: I agree, same for the two others below, there are not needed anymore I think

File: src/Symfony/Component/Cache/Adapter/ChainAdapter.php [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function getItem($key)
+    {
+        foreach ($this->adapters as $adapter) {
+            $item = $adapter->getItem($key);
+
+            if ($item->isHit()) {
+                return $item;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/ChainAdapter.php]: I'm wondering: when an item is found in a lower level adapter, shouldn't it be stored in the upper level adapters?

**Comment by nicolas-grekas** [comment]
Replaced by #18215
